llama: update llama.cpp to latest version

danbev opened the pull request:
This commit updates llama.cpp to the latest/later version.

The motivation for this is that the current version of llama.cpp is a little outdated and there have been changes to the llama.cpp API and also the model format. Currently it is not possible to use the new GGUF format and many of the available models are in this new format which can make it challenging to use this crate at the moment.

The following changes have been made:
* update llama.cpp to latest version using git submodule update --remote --merge llama.cpp

* Manually copied the generated bindings.rs file from the target directory to the src directory. Hope this was the correct thing to do.

* Updated the llm-chain-llama crate to use llama_decode instead of llm_eval which has now been deprecated.

* ~A number of TODOs have been added to the code to highlight areas that I know I need to look at.~

----

This is a work in progress but I wanted to open a draft pull request sooner rather than later to get some visibility and feedback.

Currently I've been able to successfully run the `simple example`,  `few_shot`, and `stream` examples. ~The `map_reduce_llama` example is not working as of this writing which I'll look into further~.

Juzov commented:
There's a clause ignoring MaxTokens if MaxTokens == 0 (or rather the reverse). So adding MaxTokens to be equal to MaxContextSize in the examples is redundant. If you want, and its possible, you can try to alter the option to be as big as the context window by default and remove the clause.

danbev commented:
> So adding MaxTokens to be equal to MaxContextSize in the examples is redundant.

I added a MaxBatchSize option in https://github.com/sobelio/llm-chain/pull/244/commits/452ac2c6d282a95c0c1a038432ec52490c98fd1a, and it has default value of 512 which matches the value in [llama.cpp](https://github.com/ggerganov/llama.cpp/blob/4fea3420ee3918d125d74c94d962a6ea82875351/llama.cpp#L8368-L8387), and have now removed the options from the examples (apart from simple_llama).

> If you want, and its possible, you can try to alter the option to be as big as the context window by default and remove the clause.

I'm planning on taking a closer look at the model options today, and I'll also take another look at the context options and your suggestion. Thanks

Juzov commented on crates/llm-chain-llama/src/batch.rs:
```diff
@@ -0,0 +1,136 @@
+    }
+}
+
+impl From<llama_batch> for LlamaBatch {
```
when do we use this?

danbev replied:
Actually I don't think it is used at all anymore. I'll remove it.

Juzov commented on crates/llm-chain-llama/src/executor.rs:
```diff
@@ -147,47 +157,36 @@ impl Executor {
-                if tok == tokenized_stop_prompt[stop_sequence_i] {
+                // If the prompt is in the form of a question then next
+                // predicted tok will be a new line.
+                if tok == tokenized_stop_prompt[stop_sequence_i] && n_samples > 2 {
```
I don't fully understand the n_samples > 2. If you're trying to handle another case. Could we break it up into another if case

Juzov replied:
I think I got it. But what if you actually expect a no answer?

danbev replied:
Ah good point, let me think about that and revisit this. 

@Juzov Thanks for your review and comments, I appreciate the feedback!

danbev replied:
I've taken another look at this in https://github.com/sobelio/llm-chain/pull/244/commits/d73826bd21fc761266222222382f639f3cf2f13e. I've not been able to get the llm to respond with nothing yet, but I've extracted this into a separate check like you suggested and I think this should take care of the case when an input prompt is in the form of a question.

Juzov reviewed (commented):
Looks good, added some comments. After they have been explained, we can approve this. Thanks for your effort

danbev commented on crates/llm-chain-llama/src/options.rs:
```diff
@@ -111,18 +110,58 @@ lazy_static! {
         MirostatEta: 0.1,
         PenalizeNl: true,
-        StopSequence: vec!["\n\n".to_string()]
+        StopSequence: vec!["\\n\\n".to_string()],
```
Sorry about that, it forgot to revert that change. I was struggling with the the stop sequence a little and changed this. I'll update this shortly. Thanks.

danbev commented on .github/workflows/cicd.yaml:
```diff
@@ -36,10 +36,12 @@ jobs:
 
   build_and_test:
     strategy:
+      fail-fast: false
```
I was running into an issue where the windows build would timeout which then caused the other builds to be cancelled. At that point I was not really sure if the other builds would have succeeded. I added this to allow them to proceed and not be cancelled and also later added a longer timeout to accommodate for the longer build time on windows.

danbev commented on .github/workflows/cicd.yaml:
```diff
@@ -36,10 +36,12 @@ jobs:
         os: [ubuntu-latest, macos-13, windows-latest]
         rust-version: [stable]
     runs-on: ${{ matrix.os }}
+    timeout-minutes: 30
```
Yeah, I was running into the windows build timing out and that was the reason for this change.

danbev commented on crates/llm-chain-llama/src/executor.rs:
```diff
@@ -147,47 +157,36 @@ impl Executor {
-                    let bytes_output: Vec<u8> =
-                        [leftover_bytes, context.llama_token_to_bytes(&embd[n_used])].concat();
+                    // set the batch for the next iteration.
+                    batch.set_token(tok, n_cur as i32);
```
I was mostly following what the simple example in llama.cpp does initially which is the reason. I'll have another go at this.

danbev replied:
I've updated this now so that a new LlamaBatch instance is created for each token sampled.
